### PR TITLE
Refinements to realizability

### LINKED ISSUE #4031: Good old unsoundness, CCE
http://io.livecode.ch/learn/namin/unsound/scala

```scala
object App {
  trait A { type L >: Any}
  def upcast(a: A, x: Any): a.L = x
  lazy val p: A { type L <: Nothing } = p
  def coerce(x: Any): Int = upcast(p, x)

  def main(args: Array[String]): Unit = {
    println(coerce("Uh oh!"))
  }
}
```
`lazy` above is not necessary

```scala
Exception in thread "main" java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
	at scala.runtime.BoxesRunTime.unboxToInt(BoxesRunTime.java:101)
	at App$.coerce(HelloWorld.scala:5)
	at App$.main(HelloWorld.scala:8)
	at App.main(HelloWorld.scala)
```

### PR DESCRIPTION
**Pull request by odersky**
This was another attempt to fix #4031 without breaking other things, before I found out that
it was not a fixable bug, after all. Keeping the bits that are still of value.

Supersedes #4036 and #5559.

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
Something wrong with the CI? Test output for bootstrapped after 17.06 minutes is

    Exit code 1

and nothing else.

**Comment by odersky** [comment]
Turns out, #4031 was not a bug (or, rather, not a bug that should be fixed before we track nullability). Keeping the rest of this PR since refines realizability in useful ways.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -404,8 +404,9 @@ class TypeComparer(initctx: Context) extends ConstraintHandling {
         if (cls2.isClass) {
           if (cls2.typeParams.isEmpty) {
             if (cls2 eq AnyKindClass) return true
-            if (tp1.isRef(defn.NothingClass)) return true
+            if (tp1.isRef(NothingClass)) return true
             if (tp1.isLambdaSub) return false
+            if (cls2 eq AnyClass) return true
               // Note: We would like to replace this by `if (tp1.hasHigherKind)`
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala]: This comment is now attached to the wrong line (it refers to isLambdaSub above)

**Comment by odersky** [comment]
@Blaisorblade I propose the following order to finish with realizability: You do a quick review of #5568 and we merge it. Then we rebase #5558 on top of the merged master and I review that one in turn. Sounds good?

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 1 above)
Comment (Blaisorblade) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala]: Addressed.

**Pull request merged**